resource: Add Resource[Endpoints] and adapt existing watchers

joamaki opened the pull request:
Adds `Resource[*Endpoints]` which moves the resolving of the Endpoint or EndpointSlice into the constructor for the resource. K8sWatcher refactored to use the `Resource[*Endpoints]` and code cleaned up to work with just `*Endpoints` rather than the 3 different kinds. Some messiness around collapsing them into one type and hiding the underlying resource kind. Please review those changes carefully (CachesSynced, metrics in endpoints watcher, ipcache.ResourceKindEndpoint). I've tried to keep existing behavior as much as I could, but could've potentially missed something.

See individual commits and #23734 for motivation.

joamaki committed:
resource: Add WithTransform option

Add the option to transform the object before it's added to the store.
Useful in situations where we want to store a subset of the data or
a derivate object vs the upstream object.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s: Make Endpoints object conform to runtime.Object

In preparation of making endpoints a Resource, allow transforming K8s
Endpoints/EndpointSlices into our internal representation (k8s.Endpoints)
and to store it in cache.Store, this adds the object metadata to it.

The "EndpointSliceID" is included in k8s.Endpoints so it doesn't
need to be passed separately.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s: Add Resource[*Endpoints] to shared resources

Add Resource[*Endpoints] which probes the api-server capabilities
to pick the right upstream resource kind and expose endpoints with
our internal type rather than requiring each use-site to figure
this out on their own.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
agent, operator, clustermesh-apiserver: use Resource[*Endpoints]

- Remove endpoint and endpoint_slice watchers
- Add "endpoints watcher" adapted from endpoint_slice watcher that uses
  Resource[*Endpoints] instead of informer. This keeps things the same
  downstream.
- Adapts BGP speaker to only handle *k8s.Endpoints
- Removes endpoint/endpointslice informers from operator and use resource instead

Closes: #23734
Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -458,14 +459,6 @@ func (k *K8sWatcher) resourceGroups() (beforeNodeInitGroups, afterNodeInitGroups
-	if k8s.SupportsEndpointSlice() {
-		k8sGroups = append(k8sGroups, resources.K8sAPIGroupEndpointSliceV1Beta1Discovery)
-	} else if k8s.SupportsEndpointSliceV1() {
-		k8sGroups = append(k8sGroups, resources.K8sAPIGroupEndpointSliceV1Discovery)
```
Ah probably should default to `EndpointSliceV1` for this and the metrics instead of `EndpointV1Core` as above.

YutaroHayakawa commented on pkg/k8s/watchers/endpoints.go:
```diff
@@ -0,0 +1,131 @@
+				case resource.Sync:
+					synced.Store(true)
+				case resource.Upsert:
+					k.K8sEventReceived(apiGroup, resources.MetricEndpointSlice, resources.MetricUpdate, true, false)
```
Should we use the underlying resource to manage metrics instead of blindly use EndpointSlice metrics? Same for below.

YutaroHayakawa replied:
Oh, this was covered in the PR comment. Sorry. My opinion around here is it affects the metrics label and that's a user interface we need to maintain AFAIU. I don't have any objection to consolidating them into single label, but maybe we should make it opt-in or deprecate and remove after two versions?

joamaki replied:
Changed this up now to figure out which metric name to use in `pkg/k8s/watchers/endpoints.go`. Next steps to further clean this up is to implement metrics in `Resource[T]` itself. This is mostly blocked on the metrics PR. After that we can remove the metrics from `K8sWatchers` that use `Resource[T]` and later remove `K8sWatchers` completely once we've migrated code to use the `Resource[T]` directly.

YutaroHayakawa commented on pkg/k8s/watchers/endpoints.go:
```diff
@@ -0,0 +1,131 @@
+		return
+	}
+	resource := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindEndpointSlice,
```
Is treating Endpoints as `ipcacheTypes.ResourceKindEndpointSlice` safe regardless of underlying type?

joamaki replied:
This doesn't appear to be persisted to anywhere (checked ipcachemap, kvstore and CiliumEndpoint), so appears to be safe. I've now simplified this to reduce it to just `ResourceKindEndpoint`. This would allow us to switch early to our internal endpoint representation and not carry the concrete endpoint resource kind to everywhere. @joestringer is this OK or am I potentially missing something?

giorio94 commented on pkg/k8s/endpoints.go:
```diff
@@ -129,17 +137,21 @@ func (e *Endpoints) Prefixes() []netip.Prefix {
+			Name:      ep.ObjectMeta.Name,
+			Namespace: ep.ObjectMeta.Namespace,
+		},
+		EndpointSliceName: ep.ObjectMeta.Name,
```
Nit. `EndpointSliceID` and `EndpointSliceName` could be likely changed to `EndpointID` and `EndpointName` since they are no longer epslice-specific.

joamaki replied:
Good idea. I'd propose to leave that for follow-up PR as it's a sizeable cleanup on its own.

giorio94 commented on pkg/k8s/service_cache_test.go:
```diff
@@ -118,7 +118,7 @@ func (s *K8sSuite) TestGetUniqueServiceFrontends(c *check.C) {
 
 func (s *K8sSuite) TestServiceCacheEndpoints(c *check.C) {
-	k8sEndpoints := &slim_corev1.Endpoints{
+	endpoints := ParseEndpoints(&slim_corev1.Endpoints{
```
Nit. This change (and a few others in this file) should not be part of this commit, but likely of a subsequent one.

joamaki replied:
Oops yeah, those shouldn't have been here. Moving.

giorio94 commented on pkg/k8s/watchers/watcher_test.go:
```diff
@@ -220,9 +218,9 @@ func (s *K8sWatcherSuite) TestUpdateToServiceEndpointsGH9525(c *C) {
 
 	w.K8sSvcCache.UpdateService(k8sSvc, swg)
-	w.K8sSvcCache.UpdateEndpoints(ep1stApply, swg)
+	w.K8sSvcCache.UpdateEndpoints(k8s.ParseEndpoints(ep1stApply), swg)
```
Same here (and below)

giorio94 commented on pkg/k8s/endpoints.go:
```diff
@@ -31,9 +33,15 @@ import (
+	types.UnserializableObject
+	slim_metav1.ObjectMeta
+
+	EndpointSliceID
```
Nit. The EndpointSliceName seems redundant now that Endpoints embeds ObjectMeta. I wonder if it could make sense to just embed ServiceID here (and provide an helper function to return the full ID if required).

joamaki replied:
Agreed, I'll do this in a follow-up (along with https://github.com/cilium/cilium/pull/23977#discussion_r1201798177)

giorio94 commented on operator/watchers/k8s_service_sync.go:
```diff
@@ -111,16 +102,12 @@ type ServiceSyncConfiguration interface {
-
 	readyChan := make(chan struct{}, 0)
 
+	wg.Add(1)
```
The corresponding `wg.Done()` is missing.

joamaki replied:
Fixed.

giorio94 commented on operator/watchers/k8s_service_sync.go:
```diff
@@ -129,226 +116,73 @@ func StartSynchronizingServices(ctx context.Context, wg *sync.WaitGroup, clients
+		// Wait until service cache is ready. k8sSvcCacheSynced is used by
+		// GetServiceIP() to not query incomplete service cache.
+		swg.Stop()
+		swg.Wait()
```
I'm not confident that this change preserves the previous behavior. When the `swg` is stopped, all new calls to `Add` become no-ops. Since the `swg` is immediately stopped here, once a sufficient number of events (based on the number of adds happened before this moment) will have been processed (hence calling `Done()`), `Wait` will unlock, with no guarantee that the list completed (and that all events received previously had actually been processed).

joamaki replied:
It should be correct since we're doing `swg.Add()` and then starting to handle events and doing `swg.Done()` when we receive `resource.Sync` (we care about the sync and all events handled prior to it all the way to datapath). I'll add a few comments.

giorio94 replied:
The issue is that `Add()` is a no-op once the swg is stopped, and since here it is immediately stopped, all adds performed by e.g., `K8sSvcCache.UpdateService` don't have any effect (regardless of whether before or after `resource.Sync`). Differently, all `Done()` executions always decrement the counter (until it reaches zero). Hence, when `K8sSvcCache.UpdateService` is called, Add() would not increment the counter (since the swg is stopped), but `Done()` would decrement it, making it reach 0 before the sync has actually completed.

joamaki replied:
Ah, right, I see what you mean. I'll refactor this to call `Stop` when both endpoints and services have synced so all `Add`s are effective.

giorio94 commented on pkg/k8s/watchers/resources/resources.go:
```diff
@@ -8,16 +8,11 @@ package resources
-	K8sAPIGroupEndpointSliceV1Beta1Discovery = "discovery/v1beta1::EndpointSlice"
-	// K8sAPIGroupEndpointSliceV1Beta1Discovery is the identifier for K8s resources of type discovery/v1/EndpointSlice.
-	// todo(tom): double check the uses of these two.
+	// K8sAPIGroupEndpointSliceV1Discovery is the identifier for K8s resources of type discovery/v1/EndpointSlice.
```
Nit. I would clarify that it actually represents all types of endpoint/endpointslices.

joamaki replied:
Renamed it to `K8sAPIGroupEndpointSliceOrEndpoint` to make it clear.

giorio94 commented on pkg/k8s/watchers/endpoints.go:
```diff
@@ -0,0 +1,135 @@
+	"github.com/cilium/cilium/pkg/source"
+)
+
+func (k *K8sWatcher) newEndpointsInit(swgEps *lock.StoppableWaitGroup) {
```
Nit. The swg could be initialized in this function rather than passed as parameter.

giorio94 reviewed (changes requested):
Thanks! AFAICT, this looks mostly good to me. I've left a few comments inline, mostly concerning nits.

squeed commented on pkg/k8s/endpoints.go:
```diff
@@ -31,9 +33,15 @@ import (
 // frontend ports of the corresponding service.
```
While you're here, can you add a comment indicating that this is developed from either an Endpoints or EndpointSlices object? Just so it's clear that we prefer EndpointSlices and fall back to Endpoints for very old versions.

joamaki replied:
Added comment.

squeed commented on pkg/k8s/shared_resources.go:
```diff
@@ -150,3 +161,89 @@ func ciliumCIDRGroup(lc hive.Lifecycle, cs client.Clientset) (resource.Resource[
+
+func (lw *endpointsListerWatcher) getListerWatcher() cache.ListerWatcher {
+	lw.once.Do(func() {
+		if SupportsEndpointSlice() {
```
FYI, there is a very slight behavior difference with the existing code: in addition to checking to see that the resources are known to the apiserver, it also looks to see that there is at least one EndpointSlice (presumably added by the apiserver itself).

However, I think the existing behavior is overly paranoid and safe to remove.

squeed replied:
Correspondingly, the HasEndpointSlice function can now be removed.

joamaki replied:
Thanks for spotting, removed it.

squeed commented on pkg/k8s/watchers/endpoints.go:
```diff
@@ -0,0 +1,135 @@
+	k.addKubeAPIServerServiceEndpoints(eps)
+}
+
+func (k *K8sWatcher) addKubeAPIServerServiceEndpoints(eps *k8s.Endpoints) {
```
There is an existing bug that this PR has preserved: if somehow the `default/kubernetes` service has two EndpointSlices, then we'll compute labels wrong.

This is currently not an issue, since only one EndpointSlice is created, but this lurks, waiting to trip us up.

YutaroHayakawa replied:
Does it mean when we do dual stack, that will become a problem?

squeed replied:
Funny that you mention this, because right now upstream k8s is stalled on how to announce dual-stack apiservers. But, yes, you're right.

joamaki committed:
k8s: Drop orphan function HasEndpointSlice

This was used by the EndpointSlice watcher to check whether there
really existed an EndpointSlice in addition to the API server
supporting them. Since EndpointSlices have been the default for
many versions now, we don't need this level of paranoia.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

YutaroHayakawa commented on operator/watchers/k8s_service_sync.go:
```diff
@@ -111,244 +102,106 @@ type ServiceSyncConfiguration interface {
+					k8sEventMetric(resources.MetricEndpointSlice, resources.MetricUpdate)
+					K8sSvcCache.UpdateEndpoints(ev.Object, swg)
+				case resource.Delete:
+					k8sEventMetric(resources.MetricEndpointSlice, resources.MetricDelete)
```
Maybe we need to change the metric name depending on the underlying types here?

YutaroHayakawa reviewed (approved):
Now looks good to me 👍

meyskens reviewed (approved):
LGTM for me
